vermouth/gmx/itp_read.py

```diff
@@ -244,289 +208,187 @@ def _block(self, line, lineno=0):
             context,
-            context_type=context_type,
             section=interaction_name,
-            natoms=n_atoms,
-            delete=delete,
-            current_meta=self.current_meta
+            atom_idxs=atom_idxs,
             )
 
+    def _split_atoms_and_parameters(self, tokens, atom_idxs):
+        """
+        Returns atoms from line based on the indices defined in `atom_idxs`.
+        It also interprets slices etc. stored as strings.
+
+        Parameters:
+        ------------
+        tokens: collections.deque[str]
+            Deque of token to inspect. The deque **can be modified** in place.
+        atom_idxs: list of ints or strings that are valid python slices
+
+        Returns:
+        -----------
+        list
+        """
+        def _string_to_slice(string):
+            """
+            """
+            #adopted from https://stackoverflow.com/questions/680826/
+            #python-create-slice-object-from-string
+            return slice(*map(lambda x: int(x.strip()) if x.strip() else None,
+                              string.split(':')))
+
+        atoms = []
+        remove = []
+        # first we extract the atoms from the indices given using
+        # ints or slices
+        tokens = list(tokens)
+        for idx in atom_idxs:
+            if isinstance(idx, int):
+                atoms.append([tokens[idx], {}])
+                remove.append(idx)
+            elif isinstance(idx, str):
+                _slice = _string_to_slice(idx)
+                atoms += [[atom, {}] for atom in tokens[_slice]]
+                idx_range = range(0, len(tokens))
+                remove += [i for i in idx_range[_slice]]
+            else:
+                raise IOError
+
+        # everything that is left are parameters, which we
+        # get by simply deleting the atoms from tokens
+        print(remove)
+        for index in sorted(remove, reverse=True):
+            del tokens[index]
+
+        return atoms, tokens
+
+
+    def _treat_block_interaction_atoms(self, atoms, context, section):
+        atom_names = list(context.nodes)
+        all_references = []
+        for atom in atoms:
+            reference = atom[0]
+            if reference.isdigit():
+                if int(reference) < 1:
+                    msg = ('In section {} is a negative atom reference, which is not allowed.')
+                    raise IOError(msg.format(section.name))
+
+               # The indices in the file are 1-based
+                reference = int(reference) - 1
+                try:
+                    reference = atom_names[reference]
+                except IndexError:
+                    msg = ('There are {} atoms defined in the block "{}". '
+                           'Interaction in section "{}" cannot refer to '
+                           'atom index {}.')
+                    raise IOError(msg.format(len(context), context.name,
+                                             section, reference + 1))
```

Comment:
Although formally correct (so this is a suggestion), we can make the parser a bit more lenient by not enforcing that atoms are in order and sequential, so long as their indices are unique. Also, note that here you don't check that the index is actually what you expect it is.

I would suggest looking for the node with the correct ID in `context.nodes`. Which means you may want to create a dict of `atom_id: node_idx` beforehand.